include/quic/connection.hpp

```diff
@@ -290,7 +306,13 @@ namespace oxen::quic
 
         dgram_interface di;
 
+        std::shared_ptr<void> user_data;
+
       public:
+        void set_user_data(void* data);
+
+        std::shared_ptr<void> get_user_data() { return user_data; }
```

Comment:
Should we just make the member public rather than go through a setter/getter?

Reply:
If i add virtual methods to conn interface, we can get and set from the application level. I feel like that would be useful?

Reply:
Copying from out-of-band discussion:

```C++
In connection_interface:
template <typename T, typename... Args>
void make_user_data(Args&&... args) {
    get_user_data() = std::make_shared<T>(std::forward(args)...);
}

virtual std::shared_ptr<void>& get_user_data() = 0;

template <typename T, typename = std::enable_if_t<!std::is_void<T>>>
std::shared_ptr<T> get_user_data() {
    return std::static_pointer_cast<T>(get_user_data());
}
```

Or perhaps:
```C++
protected:

virtual std::shared_ptr<void>& user_data_ptr() = 0;

public:

template <typename T, typename... Args>
void make_user_data(Args&&... args) {
    user_data_ptr() = std::make_shared<T>(std::forward(args)...);
}

public:
template <typename T, typename = std::enable_if_t<!std::is_void<T>>>
std::shared_ptr<T> get_user_data() {
    return std::static_pointer_cast<T>(user_data_ptr());
}
```

And then in Connection:
```
std::shared_ptr<void>& user_data_ptr() override { return user_data; }
```

Reply:
Thanks for copying this here!!

Reply:
Follow-up on out of band discussion:
We want to differentiate between `get_user_data` and `take_user_data` right? Since the virtual method returns a `shared_ptr` reference, calling the templated method would increase the use count to 2, making it impossible for the user to destroy the data inside `Connection` by calling `::reset()` on their `shared_ptr`.

I know the user could call `get_user_data<T>()` to steal it, but maybe we have two methods?
- `view_user_data` to access it
- `get_user_data` to take control of it